src/lib/optimizer/strategy/nested_expression_rule.cpp

```diff
@@ -0,0 +1,163 @@
+#include "nested_expression_rule.hpp"
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "constant_mappings.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/lqp_column_reference.hpp"
+#include "logical_query_plan/lqp_expression.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "resolve_type.hpp"
+#include "utils/arithmetic_operator_expression.hpp"
+
+namespace opossum {
+
+std::string NestedExpressionRule::name() const { return "Nested Expression Rule"; }
+
+bool NestedExpressionRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) {
+  if (node->type() != LQPNodeType::Projection) {
+    return _apply_to_children(node);
+  }
+
+  auto projection_node = std::dynamic_pointer_cast<ProjectionNode>(node);
+
+  const auto column_expressions = projection_node->column_expressions();
+  for (auto column_id = ColumnID{0}; column_id < column_expressions.size(); column_id++) {
+    const auto expression = column_expressions[column_id];
+    if (!expression->is_arithmetic_operator()) {
+      continue;
+    }
+
+    // If the expression contains operands with different types, or one of the operands is a column reference,
+    // it can't be resolved in this rule, and will be handled by the Projection later.
+    const auto expression_type = _get_type_of_expression(expression);
+    if (expression_type == std::nullopt) {
+      continue;
+    }
+
+    auto value = NULL_VALUE;
+    resolve_data_type(*expression_type, [&](auto type) { value = _evaluate_expression(type, expression); });
+
+    // If the value is still NULL_VALUE, then the expression could not be resolved at this point,
+    // e.g. because it is not an arithmetic operator.
```

Comment:
We should not abuse `NULL_VALUE` like that. `NULL + 3` is an expression that can be resolved at planning time.

Reply:
Yes you're right, using `nullopt` for this now.
`NULL` itself can now be resolved in this rule, as of `NULL + anything` comes out `NULL`